File: backend/app/core/domain/Plan.py
from typing import List, Set, Tuple

from app.api.schemas.schemas import PlanJSON
from app.core.domain import Problem, Profile, Topic


class Plan:
    def __init__(
        self,
        id: int,
        title: str,
        profile: Profile,
        problems: List[Tuple[bool, Problem]] = [],
        given_topics: List[Topic] = [],
    ):
        self._id = id
        self._title = title
        self._profile = profile
        self._problems = problems
        self._given_topics = given_topics
        self._topics = self._calculate_topics()

    @property
    def id(self) -> int:
        return self._id

    @property
    def title(self) -> str:
        return self._title

    @title.setter
    def title(self, value: str):
        if not value or not isinstance(value, str):
            raise ValueError("Title must be a non-empty string")
        self._title = value

    @property
    def given_topics(self) ->  List[Topic]:
        return self._given_topics

    @given_topics.setter
    def given_topics(self, value: List[Topic]):
        if not value or not isinstance(value, List[Topic]):
            raise ValueError("Given topics must be a list of topics")
        self.given_topics = value

    @property
    def profile(self) -> Profile:
        return self._profile

    @property
    def problems(self) -> List[Tuple[bool, Problem]]:
        return self._problems.copy()

    @problems.setter
    def problems(self, value: List[Tuple[bool, Problem]]) -> None:
        if not isinstance(value, list):
            raise ValueError("Problems must be a list")
        
        for item in value:
            if not isinstance(item, tuple) or len(item) != 2:
                raise ValueError("Each problem must be a (bool, Problem) tuple")
            completed, problem = item
            if not isinstance(completed, bool):
                raise ValueError("First tuple element must be a boolean")
            if not isinstance(problem, Problem):
                raise ValueError("Second tuple element must be a Problem")
        
        self._problems = value.copy()
        self._topics = self._calculate_topics()

    @property
    def topics(self) -> Set[Topic]:
        return self._topics.copy()

    def get_problem_by_id(self, problem_id: int) -> Tuple[bool, Problem]:
        problem = next(
            (p for p in self._problems if p[1].id == problem_id),
            None
        )
        if problem is None:
            raise ValueError(f"Problem {problem_id} not found in plan")
        return problem

    def toggle_problem_completion(self, problem_id: int):
        for i, (completed, problem) in enumerate(self._problems):
            if problem.id == problem_id:
                self._problems[i] = (not completed, problem)
                return
        raise ValueError(f"Problem {problem_id} not found in plan")

    def get_percentage_completed(self) -> float:
        if not self._problems:
            return 100.0
        completed = sum(1 for done, _ in self._problems if done)
        return round((completed / len(self._problems)) * 100)

    def to_json(self) -> PlanJSON:
        return PlanJSON(
            id=self._id,
            title=self._title,
            profile=self._profile.to_json(),
            problems=[(done, prob.to_json()) for done, prob in self._problems],
            topics=[t.to_json() for t in self._topics] if self._topics else None,
            given_topics=[topic.to_json() for topic in self._given_topics]
        )

    @classmethod
    def from_json(self, json_data: PlanJSON) -> 'Plan':
        try:
            if not json_data.id or not json_data.title or not json_data.profile or not json_data.given_topics:
                raise ValueError("Invalid Plan JSON - missing required fields")
            problems = [
                (bool(done), Problem.from_json(prob))
                for done, prob in json_data.problems
            ]
            given_topics = [Topic.from_json(topic) for topic in json_data.given_topics]
            plan = self(
                id=json_data.id,
                title=json_data.title,
                profile=Profile.from_json(json_data.profile),
                problems=problems,
                given_topics=given_topics
            )
            if json_data.topics:
                plan._topics = {Topic.from_json(t) for t in json_data.topics}
            return plan
        except Exception as e:
            raise ValueError(f"Invalid Plan JSON")
    
    def _calculate_topics(self) -> Set[Topic]:
        topics = set()
        for _, problem in self._problems:
            for topic in problem.topics:
                topics.add(topic)
        return topics

File: frontend/app/lib/domain/Plan.ts
import { Problem, ProblemJSON } from "./Problem";
import { Profile, ProfileJSON } from "./Profile";
import { Topic, TopicJSON } from "./Topic";

export interface PlanJSON {
  id: number;
  title: string;
  profile: ProfileJSON;
  problems: Array<[boolean, ProblemJSON]>;
  topics?: Array<TopicJSON>;
  given_topics: Array<TopicJSON>;
}

export class Plan {
  private readonly _topics: Set<Topic>;

  public constructor(
    private readonly _id: number,
    private _title: string,
    private readonly _profile: Profile,
    private _problems: Array<[boolean, Problem]> = [],
    private _given_topics: Array<Topic> = []
  ) {
    this._topics = this.calculateTopics();
  }

  // ====================
  // Public API
  // ====================

  public get id(): number {
    return this._id;
  }

  public get title(): string {
    return this._title;
  }

  public get profile(): Profile {
    return this._profile;
  }

  public get problems(): Array<[boolean, Problem]> {
    return [...this._problems] as const;
  }

  public get given_topics(): Array<Topic> {
    return [...this._given_topics] as const;
  }

  public get topics(): ReadonlySet<Topic> {
    return new Set(this._topics);
  }

  // ====================
  // Domain Methods
  // ====================

  public updateTitle(newTitle: string): void {
    this._title = newTitle;
  }

  public getProblemById(problemId: number): [boolean, Problem] | undefined {
    const result = this._problems.find(
      (problem) => problem[1].id === problemId
    );
    if (result === undefined) {
      throw new Error(`Problem ${problemId} not found in plan`);
    }
    return result;
  }

  public toggleProblemCompletion(problemId: number): void {
    const problemIndex = this._problems.findIndex(
      ([_, problem]) => problem.id === problemId
    );
    if (problemIndex === -1) {
      throw new Error(`Problem ${problemId} not found in plan`);
    }
    this._problems[problemIndex][0] = !this._problems[problemIndex][0];
  }

  public getPercentageCompleted(): number {
    if (this._problems.length === 0) return 0;
    const completedCount = this._problems.reduce(
      (count, [isDone]) => (isDone ? count + 1 : count),
      0
    );
    return Math.round((completedCount / this._problems.length) * 100);
  }

  public toJSON(): PlanJSON {
    return {
      id: this._id,
      title: this._title,
      profile: this._profile.toJSON(),
      problems: this._problems.map(([completed, problem]) => [
        completed,
        problem.toJSON(),
      ]),
      topics: Array.from(this._topics).map((topic) => topic.toJSON()),
      given_topics: Array.from(this._given_topics).map((topic) => topic.toJSON()),
    };
  }

  static fromJSON(json: PlanJSON): Plan {
    if (
      json?.id === undefined ||
      json?.id === null ||
      json?.title === undefined ||
      json?.title === null ||
      json?.profile === null ||
      json?.profile === undefined ||
      json?.given_topics === null ||
      json?.given_topics === undefined
    ) {
      throw new Error("Invalid Plan JSON");
    }
    const problems = (json.problems || []).map(([completed, problemJson]) => [
      completed,
      Problem.fromJSON(problemJson),
    ]) as Array<[boolean, Problem]>;

    const given_topics = (json.given_topics || []).map(topic_json => Topic.fromJSON(topic_json)) as Array<Topic>;
    const plan = new Plan(
      json.id,
      json.title,
      Profile.fromJSON(json.profile),
      problems,
      given_topics
    );

    if (json.topics) {
      (plan as any)._topics = new Set(
        json.topics.map((topicJson) => Topic.fromJSON(topicJson))
      );
    }

    Object.setPrototypeOf(plan, Plan.prototype);
    return plan;
  }

  public toString(): string {
    return `Plan(id:${this._id}, title:${this._title}, profile:${this._profile}, problems:${this._problems}, given_topics:${this._given_topics})`;
  }

  // ====================
  // Private Helpers
  // ====================

  private calculateTopics(): Set<Topic> {
    const topics = new Set<Topic>();
    this._problems.forEach(([_, problem]) => {
      problem.topics.forEach((topic) => topics.add(topic));
    });
    return topics;
  }
}
